### PR TITLE
Update FlorisInterface and subset of examples

### PR DESCRIPTION
**Pull request by paulf81**
# Update FlorisInterface and the non-visualization examples

This pull request updates `FlorisInterface` to support the 4-dimensional data structures along with the subset of the examples that do not require the visualization methods.

Overview of changes:
- FlorisInterface:
  - Converting all variables to be 4d with `(findex, n turbines)` as the first two dimensions
  - Remove `time_series` flag in `FlorisInterface.reinitialize`
- floris_interface_test.py
  - Correcting existing tests to work in 4D (all pass)
  - Adding additional unit test
  - Adding unit tests to check power functions:
    - test_get_turbine_powers
    - test_get_farm_power
    - test_get_farm_aep
    - test_get_farm_aep_with_conditions

### Output of Example 01

Prior to update:
```
========================= Single Wind Direction and Wind Speed =========================
The turbine power matrix should be of dimensions 1 WD X 1 WS X 2 Turbines
[[[1691.32664838  361.23224316]]]
Shape:  (1, 1, 2)

========================= Single Wind Direction and Multiple Wind Speeds ===============
The turbine power matrix should be of dimensions 1 WD X 3 WS X 2 Turbines
[[[1691.32664838  361.23224316]
  [2407.84167188  536.13924686]
  [3298.06715556  776.52510576]]]
Shape:  (1, 3, 2)

========================= Multiple Wind Directions and Multiple Wind Speeds ============
The turbine power matrix should be of dimensions 3 WD X 3 WS X 2 Turbines
[[[1691.32664838 1290.71423024]
  [2407.84167188 1837.02869145]
  [3298.06715556 2530.13309911]]

 [[1691.32664838  361.23224316]
  [2407.84167188  536.13924686]
  [3298.06715556  776.52510576]]

 [[1691.32664838 1290.71717789]
  [2407.84167188 1837.03293726]
  [3298.06715556 2530.13884175]]]
Shape:  (3, 3, 2)
```
After update:
```
========================= Single Wind Direction and Wind Speed =========================
The turbine power matrix should be of dimensions 1 FINDEX X 2 Turbines
[[1691.32664838  361.23224316]]
Shape:  (1, 2)

========================= Single Wind Direction and Multiple Wind Speeds ===============
The turbine power matrix should be of dimensions 9 FINDEX X 2 Turbines
[[1691.32664838  361.23224316]
 [2407.84167188  536.13924686]
 [3298.06715556  776.52510576]]
Shape:  (3, 2)

========================= Multiple Wind Directions and Multiple Wind Speeds ============
The turbine power matrix should be of dimensions 9 WD/WS X 2 Turbines
[[1691.32664838 1290.71423024]
 [2407.84167188 1837.02869145]
 [3298.06715556 2530.13309911]
 [1691.32664838  361.23224316]
 [2407.84167188  536.13924686]
 [3298.06715556  776.52510576]
 [1691.32664838 1290.71717789]
 [2407.84167188 1837.03293726]
 [3298.06715556 2530.13884175]]
Shape:  (9, 2)
```

### CODE REVIEW

File: examples/04_sweep_wind_directions.py [hunk 1]
```diff
@@ -22,18 +22,15 @@
...
-and the powers of the two simulated turbines is computed for all
-wind directions in one call
+This example sweeps across wind directions while holding wind speed
+constant
 
 The power of both turbines for each wind direction is then plotted
 
 """
 
 # Instantiate FLORIS using either the GCH or CC model
```
Comment (rafmudaf) [inline examples/04_sweep_wind_directions.py:32]: Update comment to reflect the line deleted below

Comment (paulf81) [reply examples/04_sweep_wind_directions.py:32]: Tried to be more clear here

**Comment by paulf81** [comment]
ok @rafmudaf  I think this one is ready for review, based on the project board I think we covered all the functions and examples we set out to (the remaining ones depend on multicp, or vis/solvers then are being updated elsewhere)

File: floris/tools/floris_interface.py [hunk 2]
```diff
@@ -760,18 +754,18 @@ def get_farm_power(
...
             turbine_weights = np.ones(
                 (
-                    self.floris.flow_field.n_wind_directions,
-                    self.floris.flow_field.n_wind_speeds,
+                    self.floris.flow_field.n_findex,
                     self.floris.farm.n_turbines
                 )
             )
+
+        #TODO: PF Confirm this is correct?
```
Comment (rafmudaf) [inline floris/tools/floris_interface.py]: Is this still an open question?

File: floris/tools/floris_interface.py [hunk 3]
```diff
@@ -859,119 +852,37 @@ def get_farm_AEP(
...
+        wind_directions = np.array(self.floris.flow_field.wind_directions, copy=True)
+        farm_power = np.zeros(self.floris.flow_field.n_findex)
 
         # Determine which wind speeds we must evaluate in floris
         conditions_to_evaluate = wind_speeds >= cut_in_wind_speed
         if cut_out_wind_speed is not None:
             conditions_to_evaluate = conditions_to_evaluate & (wind_speeds < cut_out_wind_speed)
 
         # Evaluate the conditions in floris
+        #TODO confirm this is working in new framework
```
Comment (rafmudaf) [inline floris/tools/floris_interface.py]: Is this still an open question?

File: examples/01_opening_floris_computing_power.py [hunk 4]
```diff
@@ -33,46 +33,56 @@
...
-yaw_angles = np.zeros([1,1,2]) # 1 wind direction, 1 wind speed, 2 turbines
+yaw_angles = np.zeros([1, 2])  # 1 wind direction / speed, 2 turbines
 fi.calculate_wake(yaw_angles=yaw_angles)
 
 # Get the turbine powers
-turbine_powers = fi.get_turbine_powers()/1000.
-print('The turbine power matrix should be of dimensions 1 WD X 1 WS X 2 Turbines')
+turbine_powers = fi.get_turbine_powers() / 1000.0
+
+# TODO what should we call this user/facing?
```
Comment (rafmudaf) [inline examples/01_opening_floris_computing_power.py:51]: Have you resolved this?

**Review by rafmudaf — CHANGES_REQUESTED** [review]
I found a few "TODO" comments, so just checking whether these are still open questions.

File: examples/01_opening_floris_computing_power.py (hunk 4 above)
Comment (paulf81) [reply examples/01_opening_floris_computing_power.py:51]: I was thinking to leave it as FINDEX until/unless we decide on something else?

File: floris/tools/floris_interface.py (hunk 2 above)
Comment (paulf81) [reply floris/tools/floris_interface.py]: Good catch, this is resolved, I wrote a test for it to be sure

File: floris/tools/floris_interface.py (hunk 3 above)
Comment (paulf81) [reply floris/tools/floris_interface.py]: Yes, this is resolved, wrote a test to confirm all works as expected

**Review by rafmudaf — APPROVED** [review]

File: examples/01_opening_floris_computing_power.py (hunk 4 above)
Comment (misi9170) [reply examples/01_opening_floris_computing_power.py:51]: Agreed, I think that's OK for now

File: floris/tools/floris_interface.py [hunk 5]
```diff
@@ -869,109 +857,25 @@ def get_farm_AEP(
...
-            yaw_angles=yaw_angles,
-            turbine_weights=turbine_weights,
-            no_wake=no_wake)
-
-
-        # Reset the FLORIS object to the original wind speed and directions
-        self.reinitialize(
-            wind_speeds=wind_speeds,
-            wind_directions=wind_directions
-        )
```
Comment (misi9170) [inline floris/tools/floris_interface.py:974]: @paulf81 is this removed from here in order to be reimplemented in (or aligned with) the WindRose overhaul?

File: floris/tools/floris_interface.py [hunk 6]
```diff
@@ -859,7 +846,8 @@ def get_farm_AEP(
         # Copy the full wind speed array from the floris object and initialize
         # the the farm_power variable as an empty array.
         wind_speeds = np.array(self.floris.flow_field.wind_speeds, copy=True)
-        farm_power = np.zeros((self.floris.flow_field.n_wind_directions, len(wind_speeds)))
+        wind_directions = np.array(self.floris.flow_field.wind_directions, copy=True)
+        farm_power = np.zeros(self.floris.flow_field.n_findex)
 
         # Determine which wind speeds we must evaluate in floris
         conditions_to_evaluate = wind_speeds >= cut_in_wind_speed
```
Comment (misi9170) [inline floris/tools/floris_interface.py:853]: It seems that here, we are leaning towards calling the findex the "condition". Not conclusive, but worth noting.

**Review by misi9170 — APPROVED** [review]
Looks good to me.

File: floris/tools/floris_interface.py (hunk 5 above)
Comment (rafmudaf) [reply floris/tools/floris_interface.py:974]: Yes, that was the idea

File: floris/tools/floris_interface.py (hunk 6 above)
Comment (paulf81) [reply floris/tools/floris_interface.py:853]: Yeah, I drew a blank and used condition, i think it fits well for when different findex are each a different ws/wd, but would be less descriptive is ws/wd matching row to row and changing control settings, or if everything is identical in like a time series,

**Comment by paulf81** [comment]
Ok, I think all comments addressed

Paul

From: misi9170 ***@***.***>
Reply-To: rafmudaf/floris ***@***.***>
Date: Wednesday, December 13, 2023 at 2:45 PM
To: rafmudaf/floris ***@***.***>
Cc: Paul Fleming ***@***.***>, Mention ***@***.***>
Subject: Re: [rafmudaf/floris] Update FlorisInterface and subset of examples (PR #61)


CAUTION: This email originated from outside of NREL. Do not click links or open attachments unless you recognize the sender and know the content is safe.

@misi9170 approved this pull request.

Looks fine to me.

**Pull request merged**